File: backend/src/ShopKeeper.Domain/Entities/Sale.cs
namespace ShopKeeper.Domain.Entities;

using ShopKeeper.Domain.Common;
using ShopKeeper.Domain.Enums;

public class Sale : BaseEntity, ITenantEntity
{
    public Guid BusinessId { get; set; }
    public Business Business { get; set; } = default!;

    public Guid BranchId { get; set; }
    public Branch Branch { get; set; } = default!;

    public Guid? CustomerId { get; set; }
    public Customer? Customer { get; set; }

    public string SaleNumber { get; set; } = default!;

    /// <summary>Client-generated idempotency key (e.g. a Guid minted offline, before a
    /// connection exists to ask the server for one). Null for sales created online today, but
    /// every offline-queued sale sets it, so replaying a sync after a dropped response can't
    /// create a duplicate - see CreateSaleCommandHandler's use of it.</summary>
    public Guid? ClientRequestId { get; set; }

    public Guid CashierUserId { get; set; }
    public User CashierUser { get; set; } = default!;

    /// <summary>Sum of (UnitPrice * Quantity) across items, before discount and tax.</summary>
    public decimal Subtotal { get; set; }
    public decimal DiscountAmount { get; set; }
    public decimal TaxAmount { get; set; }

    /// <summary>Subtotal - DiscountAmount + TaxAmount. What the customer actually paid.</summary>
    public decimal Total { get; set; }

    /// <summary>Sum of (UnitCost * Quantity) across items - the sale's cost of goods sold.</summary>
    public decimal TotalCost { get; set; }

    /// <summary>(Subtotal - DiscountAmount) - TotalCost. Tax is a pass-through, not profit.</summary>
    public decimal GrossProfit { get; set; }

    public SaleStatus Status { get; set; } = SaleStatus.Completed;

    public DateTimeOffset? VoidedAt { get; set; }
    public Guid? VoidedByUserId { get; set; }
    public string? VoidReason { get; set; }

    public ICollection<SaleItem> Items { get; set; } = new List<SaleItem>();
    public ICollection<Payment> Payments { get; set; } = new List<Payment>();
    public ICollection<Refund> Refunds { get; set; } = new List<Refund>();
}

File: frontend/src/types/sale.ts
export type PaymentMethod = 'Cash' | 'Card' | 'MobileMoney'

export type SaleStatus = 'Completed' | 'Voided' | 'PartiallyRefunded' | 'Refunded'

export interface SellableProduct {
  productId: string
  name: string
  sku: string
  barcode: string | null
  imageUrl: string | null
  categoryId: string | null
  sellingPrice: number
  trackInventory: boolean
  quantityOnHand: number | null
}

export interface SaleItem {
  id: string
  productId: string
  productName: string
  sku: string
  quantity: number
  unitPrice: number
  unitCost: number
  discountAmount: number
  lineRevenue: number
  lineCost: number
  lineProfit: number
  refundedQuantity: number
}

export interface Payment {
  id: string
  method: PaymentMethod
  amount: number
  referenceNumber: string | null
}

export interface Sale {
  id: string
  saleNumber: string
  branchId: string
  branchName: string
  customerId: string | null
  customerName: string | null
  cashierUserId: string
  cashierName: string
  subtotal: number
  discountAmount: number
  taxAmount: number
  total: number
  totalCost: number
  grossProfit: number
  status: SaleStatus
  voidedAt: string | null
  voidReason: string | null
  createdAt: string
  items: SaleItem[]
  payments: Payment[]
}

export interface SaleListItem {
  id: string
  saleNumber: string
  branchName: string
  cashierName: string
  total: number
  grossProfit: number
  status: SaleStatus
  itemCount: number
  createdAt: string
}

export interface SaleLineInput {
  productId: string
  quantity: number
  discountAmount: number
}

export interface SalePaymentInput {
  method: PaymentMethod
  amount: number
  referenceNumber?: string | null
}

export interface CreateSalePayload {
  branchId: string
  items: SaleLineInput[]
  discountAmount: number
  payments: SalePaymentInput[]
  customerId?: string | null
  /** Client-generated idempotency key, sent on every attempt (online or queued offline)
   * so a retry - whether from a flaky connection or a resync after reconnecting - can
   * never double-sell. See CreateSaleCommandHandler's use of it on the backend. */
  clientRequestId: string
}

/** A sale that couldn't reach the server (offline, or the request failed with a real
 * network error) and is sitting in the outbox waiting to sync. It has no server-assigned
 * id/saleNumber yet, and no cost/profit figures (SellableProduct doesn't carry cost) -
 * everything shown here is derived purely from the cart at the moment it was queued. */
export interface QueuedSale {
  queued: true
  clientRequestId: string
  branchName: string
  queuedAt: string
  items: { productId: string; productName: string; quantity: number; unitPrice: number; lineRevenue: number }[]
  subtotal: number
  discountAmount: number
  total: number
  payments: { method: PaymentMethod; amount: number }[]
}

export interface Refund {
  id: string
  refundNumber: string
  saleId: string
  saleNumber: string
  reason: string
  totalAmount: number
  createdAt: string
}
